### PR TITLE
Support for User-Level Prompts

### PR DESCRIPTION
**Pull request by seanoliver**
- New "prompts" table in public scope
- Each user can have many prompts
- BUG: Updating existing prompts has a small bug I still need to fix
- Still having this issue where the server is trying to access localStorage and throwing an error in the console, but it's not blocking us from doing the db transactions.

### CODE REVIEW

**Comment by seanoliver** [comment]
Oh I also have a small type error blocking the deployment to Vercel which I'll also fix before the final PR!

**Comment by swyxio** [comment]
exciting!!!

**Comment by seanoliver** [comment]
@swyxio ready for you!

File: app/actions.ts [hunk 1]
```diff
@@ -122,3 +118,131 @@ export async function shareChat(chat: Chat) {
 
   return payload
 }
+
+export async function getPrompts(user: User) {
+  try {
+    const { data, error } = await supabase
+      .from('prompts')
+      .select('id, prompt_name, prompt_body')
```
Comment (swyxio) [inline app/actions.ts:126]: we may want to separate prompt types later on (role prompt, style prompt, environment prompt, etc) but not a dealbreaker

File: app/actions.ts [hunk 2]
```diff
@@ -122,3 +118,131 @@ export async function shareChat(chat: Chat) {
...
+      return result
+    }, {} as { [key: string]: string })
+
+
+    // Un-flatten the prompt data
+    let promptGroups: PromptGroups = {}
+
+    for (let key in promptData) {
+      const [tempField, index] = key.split('_').slice(1)
+      const field: 'id' | 'name' | 'body' = tempField as 'id' | 'name' | 'body'
```
Comment (swyxio) [inline app/actions.ts:179]: great to get this logic done now

File: app/actions.ts [hunk 3]
```diff
@@ -122,3 +118,131 @@ export async function shareChat(chat: Chat) {
...
+
+      if (!promptGroups[index]) {
+        promptGroups[index] = {}
+      }
+
+      promptGroups[index][field] = promptData[key]
+    }
+    console.log('promptGroups', promptGroups)
```
Comment (swyxio) [inline app/actions.ts:187]:
```suggestion
    // console.log('promptGroups', promptGroups)
```

File: app/actions.ts [hunk 4]
```diff
@@ -122,3 +118,131 @@ export async function shareChat(chat: Chat) {
...
+    for (let index in promptGroups) {
+      let prompt = promptGroups[index]
+      if (prompt.id) {
+        const { data, error } = await supabase
+          .from('prompts')
+          .update({
+            prompt_name: prompt.name,
```
Comment (swyxio) [inline app/actions.ts:194]: i was thinking if we can make a visual distinction as well, like add a prompt color to this so people know which prompt theyre in without reading

File: app/actions.ts [hunk 5]
```diff
@@ -122,3 +118,131 @@ export async function shareChat(chat: Chat) {
...
+        .update({ email: userData.email })
+        .eq('id', user.id)
+    }
+
+    if (userData.username) {
+      await supabase
+        .from('auth.users')
+        .update({ user_name: userData.username })
+        .eq('id', user.id)
+    }
```
Comment (swyxio) [inline app/actions.ts:231]: todo: make it harder to change username in future, ask for confirm

File: app/actions.ts [hunk 6]
```diff
@@ -122,3 +118,131 @@ export async function shareChat(chat: Chat) {
...
+  }
+}
+
+export async function updateUser({
+  values,
+  user
+}: {
+  values: { [x: string]: any }
+  user: User
+}) {
```
Comment (swyxio) [inline app/actions.ts:157]: i wonder if you can send in a curl request and just inject/overwrite whatever user you want. aka how secure is this

File: components/profile-form.tsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  user: any
+  prompts: any[]
+}) {
+
+  let formSchema: z.ZodRawShape = {
+    username: z
+      .string()
+      .min(2, { message: 'Username cannot have fewer than 2 characters' })
+      .max(50, { message: 'Username cannot have more than 50 characters' }),
+    email: z.string().email({ message: 'Please enter a valid email address' })
```
Comment (swyxio) [inline components/profile-form.tsx:35]: nice

File: components/ui/form.tsx [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+const FormFieldContext = React.createContext<FormFieldContextValue>(
+  {} as FormFieldContextValue
+)
+
+const FormField = <
+  TFieldValues extends FieldValues = FieldValues,
+  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
+>({
+  ...props
+}: ControllerProps<TFieldValues, TName>) => {
```
Comment (swyxio) [inline components/ui/form.tsx:34]: ouch haha

Comment (swyxio) [reply components/ui/form.tsx:34]: if you ever run into any trouble with form types feel free to ts ignore these things

File: components/user-menu.tsx [hunk 9]
```diff
@@ -66,7 +68,7 @@ export function UserMenu({ user }: UserMenuProps) {
           </Button>
         </DropdownMenuTrigger>
         <DropdownMenuContent sideOffset={8} align="start" className="w-[180px]">
-          <DropdownMenuItem className="flex-col items-start">
+          <DropdownMenuItem onClick={() => router.push('/profile')} className="flex-col items-start">
             <div className="text-xs font-medium">
               {user?.user_metadata.name}
             </div>
```
Comment (swyxio) [inline components/user-menu.tsx:74]: make it more obvious that there is a profile page to edit, clicking this is not enough

```suggestion
          <DropdownMenuItem onClick={() => router.push('/profile')} className="flex-col items-start cursor: pointer">
	          <div className="text-xs font-medium">
	            Profile
	          </div>
          </DropdownMenuItem>
          <DropdownMenuItem className="flex-col items-start cursor: pointer">
            <div className="text-xs font-medium">
              {user?.user_metadata.name}
            </div>
```

**Review by swyxio — APPROVED** [review]
all looks good just left some comments for consideration

File: app/actions.ts (hunk 1 above)
Comment (seanoliver) [reply app/actions.ts:126]: Ah yes, we could add a "Type" or "Category" field as an additional migration to this table. Also "Color" per your other comment.

File: app/actions.ts (hunk 6 above)
Comment (seanoliver) [reply app/actions.ts:157]: I'll make a separate note to test sending requests to it just to confirm but I believe you'd need to pass a valid JWT, and even then there's RLS so you would only be able to modify rows associated with that user.

File: components/ui/form.tsx (hunk 8 above)
Comment (seanoliver) [reply components/ui/form.tsx:34]: Sounds good haha. They've been helpful so far though.

File: app/actions.ts (hunk 6 above)
Comment (swyxio) [reply app/actions.ts:157]: true true

**Pull request merged**